build/gen.go

```diff
@@ -0,0 +1,114 @@
+//go:build ignore
+
+// This script generates meta.go which contains version info for the walletd binary. It can be run with `go generate`.
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"os"
+	"os/exec"
+	"strings"
+	"text/template"
+	"time"
+)
+
+const logFormat = `{%n  "commit": "%H",%n  "shortCommit": "%h",%n  "timestamp": "%cD",%n  "tag": "%(describe:tags=true)"%n}`
+
+type (
+	gitTime time.Time
+
+	gitMeta struct {
+		Commit      string  `json:"commit"`
+		ShortCommit string  `json:"shortCommit"`
+		Timestamp   gitTime `json:"timestamp"`
+		Tag         string  `json:"tag"`
+	}
+)
+
+var buildTemplate = template.Must(template.New("").Parse(`// Code generated by go generate; DO NOT EDIT.
+// This file was generated by go generate at {{ .RunTime }}.
+package build
+
+const (
+	commit = "{{ .Commit }}"
+	version = "{{ .Version }}"
+	buildTime = {{ .UnixTimestamp }}
+)
+`))
+
+// UnmarshalJSON implements the json.Unmarshaler interface.
+func (t *gitTime) UnmarshalJSON(buf []byte) error {
+	timeFormats := []string{
+		time.RFC1123Z,
+		"Mon, 2 Jan 2006 15:04:05 -0700",
+		"2006-01-02 15:04:05 -0700",
+		time.UnixDate,
+		time.ANSIC,
+		time.RFC3339,
+		time.RFC1123,
+	}
+
+	for _, format := range timeFormats {
+		parsed, err := time.Parse(format, strings.Trim(string(buf), `"`))
+		if err == nil {
+			*t = gitTime(parsed)
+			return nil
+		}
+	}
+	return errors.New("failed to parse time")
+}
+
+func getGitMeta() (meta gitMeta, _ error) {
+	cmd := exec.Command("git", "log", "-1", "--pretty=format:"+logFormat+"")
```

Comment:
as of Go 1.18 this is accessible without needing to shell out:

```go
import "runtime/debug"

var Commit = func() string {
    if info, ok := debug.ReadBuildInfo(); ok {
        for _, setting := range info.Settings {
            if setting.Key == "vcs.revision" {
                return setting.Value
            }
        }
    }
    return "?"
}()
```
Apparently you can get the tag as well with `info.Main.Version`, but I haven't tested that myself.
Would be nice if we could avoid having to call `go generate`.